### DISCUSSION TITLE
How to expose this DAG to port 5000 for inference

### DISCUSSION THREAD

**Discussion by rajtilakjee** [apache/airflow#34663] [Q&A] [answered]
Hello everyone, was wondering if someone can help me with this. I want to create a simple ML pipeline that would run on Airflow and Docker and listen on port 5000. Anyone can then ping on port 5000 using the curl command to get a prediction. Here's what I did so far:

Created a root folder and inside the root folder created 4 files: Dockerfile, docker-compose.yml, requirements.txt, and DAG file named pipeline_dag.py

Here are the respective content of each file. First, the Dockerfile

```
FROM apache/airflow:latest

USER airflow

COPY requirements.txt /

RUN pip install --no-cache-dir "apache-airflow==${AIRFLOW_VERSION}" -r /requirements.txt
```

Next, we have the docker-compose.yml file

```
version: '3'

services:
  sleek-airflow:
    image: pythonairflow:latest

    volumes:
      - ./airflow:/opt/airflow

    ports:
      - "8080:8080"

    command: airflow standalone
```

Then, we have the requirements.txt file

`scikit-learn`

Finally, its the DAG file

```
from airflow import DAG
from airflow.operators.python import PythonOperator
from airflow.utils.dates import days_ago
from datetime import datetime
import requests

def train():
    # Import necessary libraries
    from sklearn.datasets import fetch_california_housing
    from sklearn.model_selection import train_test_split
    from sklearn.preprocessing import StandardScaler
    from sklearn.linear_model import LinearRegression
    from sklearn.metrics import mean_squared_error

    # Step 1: Fetch the California housing dataset
    data = fetch_california_housing()

    # Step 2: Split the data into features (X) and target (y)
    X = data.data
    y = data.target

    # Step 3: Split the data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Step 4: Preprocess the data using StandardScaler
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)

    # Step 5: Prepare the model using Linear Regression
    model = LinearRegression()

    # Step 6: Train the model on the training data
    model.fit(X_train_scaled, y_train)

    # Step 7: Use the trained model for prediction
    y_pred = model.predict(X_test_scaled)

    # Step 8: Evaluate the model (e.g., calculate Mean Squared Error)
    mse = mean_squared_error(y_test, y_pred)
    print(f"Mean Squared Error: {mse}")


dag = DAG(
    'pipeline_dag',
    default_args={'start_date': days_ago(1)},
    schedule_interval='0 23 * * *',
    catchup=False
)

pipeline_task = PythonOperator(
    task_id='train_model',
    python_callable=train,
    dag=dag
)

pipeline_task
```

This is what I am doing (I am using VS Code). First, I right-click on Dockerfile and click on Build. Next, I right-click on docker-compose.yml and click on Compose Up. Now I go to Docker Desktop and find that container, view it. Once its up and running, I put the DAG file inside the airflow/dags folder, restart the container from Docker Desktop, and login to the web UI once again. Next, I run the DAG from the UI. It runs fine. However, I have never worked with Airflow before. So I am getting all confused with how to implement the rest of the requirement so that I can expose this model on port 5000 and it can listen for the curl command. Please help!

**Comment by Taragolis** [comment, marked as answer]
> I want to create a simple ML pipeline that would run on Airflow and Docker and listen on port 5000. Anyone can then ping on port 5000 using the curl command to get a prediction

That is not how Airflow work. Airflow it is workflow manager (or if simplified "smart distributed cron on steroids") and not an Application Server.

You could run DAG from [REST API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/post_dag_run) but it only return information is it started or not and then you might try to get XCom ([Documentation](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/xcoms.html), [REST API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/get_xcom_entry)) value which send by your task but in general this pipeline looks so complicated and may not suits for your requirement and expectations.
